Add ImportPSModulesFromPath support

beargle opened the pull request:
Add `Start-RSJob` parameter `ModulesFromPathToImport` that accepts a single path, from which module(s) will be imported into background runspace jobs. Modules in this path are not required to be listed by name, and the path does not have to exist in `$env:PSModulePath`. Fixes #177.

**Changes proposed in this pull request:**
 - Import module(s) into runspace job from a path specified by `Start-RSJob` parameter `-ModulesFromPathToImport`

**How to test this code:**
The snippet below allows manual verification of `Get-Module` output when either a full or relative module path name is passed. I'm relatively new to Pester, but would like to learn if someone has suggestions on how to setup the automated test cases.

```PowerShell
# assuming PoshRSJob is in $env:PSModulePath or imported, setup a sample custom module path.
# get both a full and relative path name for testing
$ModulePath = New-Item -Path ([Guid]::NewGuid().ToString()) -ItemType Directory
$RelativeModulePath = Resolve-Path -Path $ModulePath -Relative

# export sample modules for import in runspace
svn export https://github.com/psake/psake/tags/v4.7.0/src $ModulePath\psake | Out-Null;
svn export https://github.com/ramblingcookiemonster/PSDeploy/trunk/PSDeploy $ModulePath\PSDeploy | Out-Null;

# start single runspace job, passing full name of module path.
# Get-Module output in the runspace should show imported psake and PSDeploy modules.
"Modules from full path" | Start-RSJob -Name { $_ } -ModulesFromPathToImport $ModulePath -ScriptBlock {
    Get-Module;
} | Wait-RSJob | Receive-RSJob;

# do it again, this time with a relative module path
"Modules from relative path" | Start-RSJob -Name { $_ } -ModulesFromPathToImport $RelativeModulePath -Verbose -ScriptBlock {
    Get-Module;
} | Wait-RSJob | Receive-RSJob;

# cleanup
Remove-Item -Recurse -Force $ModulePath;
Remove-Variable -Name @("ModulePath", "RelativeModulePath");
```

**Has been tested on (remove any that don't apply):**
 - PowerShell 4.0
 - Windows 7 and 8.1

beargle committed:
Add support to import module(s) from custom path.

Add parameter `ModulesFromPathToImport` that accepts a single path, from
which module(s) will be imported into background runspace jobs. Modules
in this path are not required to be listed by name, and the path does
not have to exist in `$env:PSModulePath`.

beargle committed:
Validation of custom module(s) path to import.

When the '-ModulesFromPathToImport' parameter is passed for
'Start-RSJob', it should handle either a full and relative path name.

Also raise an exception if the module(s) path is not found, which means
none of the module(s) are imported in the runspace job. An exception
prevents the runspace jobs from starting, as any dependent functions
referenced inside the 'ScriptBlock' parameter would not be found.

beargle committed:
Fix exception handling in module path validation.

`Resolve-Path` fires a non-terminating error when it can't find a
specified path, which means the error isn't caught for handling
in a Catch block. Make `Resolve-Path` raise a terminating error,
so we can re-throw to add context to the original exception message.

proxb commented:
@beargle Any chance you could look at using the method shown here: https://github.com/proxb/PoshRSJob/issues/177#issuecomment-371038089? It would be nice to keep all module importing down to a single parameter if possible. If it doesn't work out though, I can move forward to accept this PR.

beargle commented:
@proxb Sure, I saw the [comment](https://github.com/proxb/PoshRSJob/issues/177#issuecomment-371038089), but am not sold on readability of pre-processing module paths before passing to `Start-RSJob`.

It seems more intuitive to be able to pass a single "root" module path, so that everything (properly organized as importable, of course) gets loaded into the runspace jobs.

I totally get the usability issue with having two parameters that could be used for module import...do you think it would be confusing to wrap functionality of both [InitialSessionState.ImportPSModule](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.runspaces.initialsessionstate.importpsmodule?view=powershellsdk-1.1.0) and [InitialSessionState.ImportPSModulesFromPath](https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.runspaces.initialsessionstate.importpsmodulesfrompath?view=powershellsdk-1.1.0) into the single `-ModulesToImport` parameter?

proxb commented:
> do you think it would be confusing to wrap functionality of both InitialSessionState.ImportPSModule and InitialSessionState.ImportPSModulesFromPath into the single -ModulesToImport parameter?

@beargle I think that would be a better way to go at least to support both the module name and module path requirement even though it will require a little more work to handle both of the possible values being supplied to the parameter.  While I get the ease of just supplying a single root math containing all of the modules, I'm just not sure about having multiple module import parameters.

beargle committed:
Consolidate module import into single parameter.

Remove separate '-ModulesFromPathToImport' parameter, adding this
functionality to existing '-ModulesToImport' parameters. This means
'-ModulesToImport' can accept a collection containing:

 - Installed module names
 - Path (full or relative) containing one or more modules
 - Path (full or relative) to a module manifest, script module, or
   binary module file

Purpose of consolidation is to simplify module import for the caller;
They should have the flexibility to pass a root module path for
recursive import, or be able to specify modules similar to `Import-Module`,
with minimal preprocessing of the parameter collection.

beargle commented:
@proxb Makes sense, I've consolidated the module path import into existing `-ModulesToImport` parameter, and updated the original PR text. It should accept these different ways of specifying modules for import:

- Installed module names
```PowerShell
Start-RSJob -ModulesToImport "PSDeploy"...
```
- Path (full or relative) to a module manifest, script module, or binary module file
```PowerShell
Start-RSJob -ModulesToImport ".\PSDeploy\PSDeploy.psd1"...
```
- Path (full or relative) containing one or more modules
```PowerShell
Start-RSJob -ModulesToImport ".\workspace\Modules"...
```
- Combination of any of the above
```PowerShell
Start-RSJob -ModulesToImport "PSDeploy", ".\workspace\Modules"...
```

Let me know your thoughts, thanks!
